### PR TITLE
Architecture modularisation

### PR DESCRIPTION
**Pull request by MrM0nkey**
In order to improve the architecture of this framework #13, here is a first step in this direction.
This pull request provides base classes for simulations and learner. With this pull request it is possible to easily extend the framework with new simulations and learner. This pull request also includes some new input transformations for the ltf array simulation.

### CODE REVIEW

File: pypuf/learner/lr.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+            return array([
+                0
+                if max[i] == neighbor[i] else
+                combined_model_responses[i] / max[i]
+                for i in range(self.training_set.N)
+            ])
+        # in a multiprocessing scenario the object references would not be the same!
+        if compare_functions(self.combiner, LTFArray.combiner_xor):
+            model_gradient = model_gradient_xor
+        elif compare_functions(self.combiner == LTFArray.combiner_ip_mod2):
```
Comment (nils-wisiol) [inline pypuf/learner/lr.py]: typo? Here we have "==" instead of the "," in line 138

File: pypuf/learner/lr.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+                  ' => ' +
+                  ('+1' if self.set.responses[idx] == 1 else '-1') +
+                  '    ' +
+                  '  '.join([('%+ 7.2f' % mr) for mr in model.ltf_eval([c])[0]]) +
+                  '    ' +
+                  ('% 10.5f' % self.class_probability[idx])
+                  )
+
+    def __getstate__(self):
+        print(self.__dict__)
```
Comment (nils-wisiol) [inline pypuf/learner/lr.py]: GitHub doesn't seem happy with no newline at the end of the file. Is that somehow important?

**Comment by MrM0nkey** [comment]
- discuss package/module names
Learner and simulations are now placed in extra packages regarding learner/simulation toplevel package.
- restructured code is added (LR class, LTFArray class), but old code is not removed
Now it is done at the specific commit,
- python 3.6 testing should be added back (was removed by earlier PR)
Added also 3.5.
- commits need to be cleaned up (in particular, the "around" commit also works on the renamed classes)
I hope it is enough.
- why renamed LogisticRegression into LR?
That was a mistake.

**Comment by MrM0nkey** [comment]
- discuss package/module names
Learner and simulations are now placed in extra packages regarding learner/simulation toplevel package.
- restructured code is added (LR class, LTFArray class), but old code is not removed
Now it is done at the specific commit,
- python 3.6 testing should be added back (was removed by earlier PR)
Added also 3.5.
- commits need to be cleaned up (in particular, the "around" commit also works on the renamed classes)
I hope it is enough.
- why renamed LogisticRegression into LR?
That was a mistake.

**Pull request merged**